api/desecapi/views.py

```diff
@@ -127,52 +130,37 @@ class RRsetDetail(generics.RetrieveUpdateDestroyAPIView):
     serializer_class = RRsetSerializer
     permission_classes = (permissions.IsAuthenticated, IsDomainOwner,)
 
-    def get_queryset(self):
-        name = self.kwargs['name']
-        subname = self.kwargs['subname']
-        type = self.kwargs['type']
-
-        return RRset.objects.filter(domain__name=name, subname=subname, type=type)
-
     def delete(self, request, *args, **kwargs):
         try:
-            return super().delete(request, *args, **kwargs)
-        except django.core.exceptions.ValidationError as e:
-            ex = ValidationError(detail={"detail": str(e)})
-            ex.status_code = status.HTTP_409_CONFLICT
-            raise ex
+            super().delete(request, *args, **kwargs)
+        except Http404:
+            pass
+        return Response(status=status.HTTP_204_NO_CONTENT)
 
-    def get(self, request, *args, **kwargs):
+    def get_queryset(self):
+        name = self.kwargs['name']
         subname = self.kwargs['subname']
         type = self.kwargs['type'].upper()
 
-        # TODO Do we need this functionality at all? We could just disregard the SOA record completely.
-        if type == 'SOA' and subname == '':
-            name = self.kwargs['name']
+        if type == 'SOA':
+            raise PermissionDenied("You cannot tamper with the SOA record.")
```

Comment:
This is more difficult than it seems. Explanation:

- Object permission operates on objects (not on queries etc.), so they only apply to objects that have actually been saved to the database. However, RRsets of restricted type are not expected to be in the database.
- I considered `type` field validation in the serializer as an alternative. However, serializer-based validation only works for user input (on `POST`); if we want to restrict `GET` requests for restricted RRset types (even if they are not present in the database), the serializer will not be involved as the restriction is before the serialization step.
- Thus, of the three "can't tamper with restricted record types" exceptions, only the one in the view's `create()` method (dealing with `POST`) can be replaced by a field validator in the serializer; the other two exceptions cannot be moved.
- An alternative _may_ be model permissions (instead of object permissions), see http://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions. Conceptually, however, this sort of implies that there are users which have "restricted types access permission" and others that don't. AFAIU, this would also create an extra entry in the `auth_permission` database table. I'm not sure this solution is desirable.

All in all, I favor doing nothing about this at this point (not even moving the single check for which that is possible). This gives the advantage of having all checks in more or less one place (namely, in the `RRset`-related views).